storage/remote/read.go

```diff
@@ -44,7 +44,7 @@ type querier struct {
 // Select implements storage.Querier and uses the given matchers to read series
 // sets from the Client.
 func (q *querier) Select(p *storage.SelectParams, matchers ...*labels.Matcher) (storage.SeriesSet, error) {
-	query, err := ToQuery(q.mint, q.maxt, matchers, p)
+	query, err := ToQuery(p.Start, p.End, matchers, p)
```

Comment:
This needs to stay as is, so that the protocol doesn't change and the remote endpoint still knows the full range of the query.

Reply:
This is actually a bug, as described in detail (with example code showing the problem) in https://github.com/prometheus/prometheus/pull/4226#issuecomment-397382955 With this the way it was we are *severely* over fetching data for promql queries that have offsets.

Reply:
I understand what you are saying and have since your initial report. You can use the new hints to make your software deal better with this scenario, but you don't get to make breaking changes.